### DISCUSSION TITLE
How to use conjoon...how to config ?

### DISCUSSION THREAD

**Discussion by SupernovaCooper** [conjoon/conjoon#34] [Q&A]
Hi, could you please help me out ? thanks.
How to run conjoon on IIS ?
What I want is run it on IIS, and then I want to add all my email accounts into it (or any kind of webmail system)... but I am not sure if conjoon is the right option for me...
I tried use the release from conjoon, and put it into IIS website, then I can visit the conjoon, but I don't know how to set, it's allow me to login with any username and any password, and after login,  I can not find how to add new email account.
sorry for the trouble, I really don't understand how to use conjoon.

I use these commonds(windows server 2019 with nodejs installed):
npx create-conjoon@latest 
npm run stage

but it ask me to install webpack-dev-server, when I try to install webpack-dev-server, it ask me to install VS2013.... why should I install such a big thing on the server...

Please help...
What I want is just use any kind webmail system, add all my different email accounts, and login to the system with my own username and password(not any email account, because I enabled  two-factor authentication, and apps password, I can not remember any apps password). which means this system is only used for myself. 
Sorry for my pool english, I am not sure if I explained clearly...
Plase help..Thanks a lot.

**Comment by ThorstenSuckow** [comment]
Hey there,

what you need to do is to install the frontend and make sure a backend is available that conforms to [rest-api-email](https://conjoon.stoplight.io/docs/rest-api-description/).

#### Frontend
Since you do not intend to develop with conjoon, I suggest to follow the  [Quick Start Guide](https://www.conjoon.org/docs/#quick-start) for dropping the frontend into a folder that will be served as the doc root for your IIS instance.

When using `create-conjoon`, make sure to select `release` when prompted for the installation type ([refer to its docs for more information](https://www.conjoon.org/docs/api/misc/@conjoon/create-conjoon)).

#### Backend
Once you have installed the client, you need to install the **rest-api-email**-backend. There is a Laravel-application that implements the API: [lumen-app-email](https://www.conjoon.org/docs/api/backends/@conjoon/lumen-app-email). It is written in PHP so if you're IIS is configured to run PHP, you're good to go.

Once you have installed both apps, you can configure conjoon to be run with multiple IMAP accounts, which can be done [by updating its configuration files](https://www.conjoon.org/docs/api/packages/@conjoon/extjs-app-localmailaccount). Let me know if you need more help!

Bests

Thorsten

**Comment by SupernovaCooper** [comment]
Hi,
Thank you for the details.
But I met error while I install the Frontend..

PS C:\inetpub\wwwroot\conjoon> npx create-conjoon@latest

   

> ╭──────────────────────────────────────────────────────────────────────────────────────────────────────────╮
>    │                                                                                                          │
>    │                                                  __                                                      │
>    │                          ___     ___     ___    /\_\     ___     ___     ___                             │
>    │                         /'___\  / __`\ /' _ `\  \/\ \   / __`\  / __`\ /' _ `\                           │
>    │                        /\ \__/ /\ \L\ \/\ \/\ \  \ \ \ /\ \L\ \/\ \L\ \/\ \/\ \                          │
>    │                        \ \____\\ \____/\ \_\ \_\ _\ \ \\ \____/\ \____/\ \_\ \_\                         │
>    │                         \/____/ \/___/  \/_/\/_//\ \_\ \\/___/  \/___/  \/_/\/_/                         │
>    │                             Create conjoon apps    \ \____/ create-conjoon@1.0.0                         │
>    │                                       easily     \/___/  https://conjoon.org                             │
>    │                                                                                                          │
>    │                                                                                                          │
>    ╰──────────────────────────────────────────────────────────────────────────────────────────────────────────╯
> 
> √ Please select the installation type. » pre-built release (release)
> [INFO] Using "release" for installation.
> √ Please specify the target folder for this installation ... ./conjoon
> [INFO] Fetching available versions of conjoon ...
> [INFO] Latest release of conjoon is 1.2.1
> √ Which version should be used for this installation? » 1.2.1
> [INFO] Using external initializer, resolving paths...
> √ What should we name this installation? ... conjoon
> [INFO] Downloading release: v1.2.1
> [INFO] Extracted to "C:\inetpub\wwwroot\conjoon\conjoon" successfully
> [SUCCESS]
>         Created conjoon.
>         Inside that directory, you will find an official release build of conjoon.
> 
>         Make sure to read the documentation at https://conjoon.org.
> 
>         Happy messaging!
> 
> PS C:\inetpub\wwwroot\conjoon> ls
> 
> 
>     目录: C:\inetpub\wwwroot\conjoon
> 
> 
> Mode                LastWriteTime         Length Name
> ----                -------------         ------ ----
> d-----  23.05.29 星期一      8:57                conjoon
> 
> 
> PS C:\inetpub\wwwroot\conjoon> cd conjoon
> PS C:\inetpub\wwwroot\conjoon\conjoon> npm run stage
> npm ERR! code ENOENT
> npm ERR! syscall open
> npm ERR! path C:\inetpub\wwwroot\conjoon\conjoon/package.json
> npm ERR! errno -4058
> npm ERR! enoent ENOENT: no such file or directory, open 'C:\inetpub\wwwroot\conjoon\conjoon\package.json'
> npm ERR! enoent This is related to npm not being able to find a file.
> npm ERR! enoent
> 
> npm ERR! A complete log of this run can be found in: C:\node.js\node_cache\_logs\2023-05-29T00_58_09_592Z-debug-0.log
> PS C:\inetpub\wwwroot\conjoon\conjoon>


It's said can not find package.json... 
So , what should I do now....
Thank you.

**Comment by SupernovaCooper** [comment]
And here is my node and npm version
PS C:\inetpub\wwwroot\conjoon\conjoon> node -v
v18.16.0
PS C:\inetpub\wwwroot\conjoon\conjoon> npm -v
9.6.7

**Reply by ThorstenSuckow** [reply]
Allright, now I know where the confusion comes from: No need for `npm run stage`, that would be for dev-versions that are served locally from your computer. When choosing the install-type `release`, the downloaded files represent a build that can be served by any webserver. In your case, you should put them in the doc-root of your IIS instance.

**Comment by SupernovaCooper** [comment]
Hi,
Sorry to bother again.

My root dir is : C:\inetpub\wwwroot\conjoon
then I use this command to create conjoon
npx create-conjoon@latest
The root of conjoon become : C:\inetpub\wwwroot\conjoon\conjoon
then I add website in IIS, and visit through 127.0.0.1:8033 or a local customer domain, seems the frontend works.

Then I use this command to install backend.
composer create-project conjoon/lumen-app-email htdocs "1.*" 
The dir of backend is:C:\inetpub\wwwroot\conjoon\htdocs
Then I add it into phpEnv(I use phpEnv to run the php website...), and visit through 127.0.0.1:8034 or local customer domain, and it shows phpinfo, seems also works...

Then what should I do...the last step "updating its configuration files. " I don't know what to do exactlly...I even can not find any file named "workspace.json"..
Sorry... I am really stupid for this...

**Reply by ThorstenSuckow** [reply]
Don't worry about it, I'm happy to help and questions like this show me where the docs have to be more verbose.

In the folder of the frontend, you will find the configuration file of the email client `desktop/resources/coon-js/conjoon.conf.json``.

You need to do the following:

### 1. Enable / disable packages
in the packages-section, you will have to disable `extjs-app-imapuser` and enable `extjs-app-localmailaccount`:

```json
 "extjs-app-localmailaccount": {
     "enabled": true,
...
 "extjs-app-imapuser": {
     "enabled": false,
...
```

### 2. Configure API-endpoints for the client
For the package `extjs-app-webmail`, you have to configure the entry `service.rest-api-email` so that the address represents the url where the `rest-api-email`-endpoints - in your case given by the **lumen-app-email**-backend - can be found (that is, your IIS PHP installation):

Example:

```json
          "extjs-app-webmail": {
               "config": {
                   ...
                    "service": {
                        "rest-api-email": {
                            "base": "https://your.iss-ip-address-or-domain-name.com/lumen-app-email-dir/app/public/rest-api-email/v0/"
                        }
                    }
                }
            },
  ...
```

### 3. Configure lumen-app-email-instance

The config file `.env` of your `lumen-app-email`-installation should look similar to this (depending on your server's IP and such), but make sure that `AUTH_PROVIDER` is set to `local-mail-account`:

```env
APP_URL=https://your.iss-ip-address-or-domain-name.com/lumen-app-email-dir/app/public
APP_EMAIL_PATH=rest-api-email
AUTH_PROVIDER=local-mail-account
APP_ENV=production
APP_DEBUG=true
```

**Reply by SupernovaCooper** [reply]
Hi,
Can I just put the frontend and backend into one folder as one PHP website ? maybe it's easy to setup...

And I can not find `desktop/resources/coon-js/conjoon.conf.json` , but found  `desktop/resources/conjoon.conf.json` and   `desktop/resources/conjoon.dev.conf.json` and `resources/conjoon.conf.json` and `resources/conjoon.dev.conf.json`.

**Reply by ThorstenSuckow** [reply]
My bad, the directory structure was flattened in a recent release so `desktop/resources/conjoon.conf.json` is the way to go.

Also no problems to use one folder for the whole app, however backend and frontend should be put into individual folders as it was meant to be (might spare you configuration issues):

```
doc_root/
   frontend/ <- conjoon goes here
   backend/ <- lumen-app-email goes here
```

The naming is up to you :)

**Reply by SupernovaCooper** [reply]
Hi,
Thanks for your patience.

1. In the config file `desktop/resources/conjoon.conf.json`, extjs-app-localmailaccount enable set to true, and extjs-app-imapuser enable set to false.

2. "rest-api-email" :
"base": "http://127.0.0.1:8056/app/public/rest-api-email/v0/"

3. For the .env file of lumen-app-email, I set APP_URL into `http://127.0.0.1:8056/app/public/`
and then I visit the backend:http://127.0.0.1:8056/app/public/ it shows:"Sorry, the page you are looking for could not be found."
when I visit `http://127.0.0.1:8056/app/public/rest-api-email/v0/` nginx shows 404

4. Finally I visit the frontend: http://127.0.0.1:8055/ it ask me to add email account. After I click create, it will shows the detail of the email account, but when I click "save" nothing happends.

5. I open the devtools window on the browser, and visit frontend http://127.0.0.1:8055/ again, and I found 3 errors from console:

A.HEAD http://127.0.0.1:8055/desktop/resources/conjoon.prod.conf.json net::ERR_ABORTED 404 (Not Found)
B.Access to XMLHttpRequest at 'http://127.0.0.1:8056/app/public/rest-api-email/v0/MailAccounts/cn_localmailaccount-1/MailFolders?_dc=1685500140332' from origin 'http://127.0.0.1:8055' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.
C.GET http://127.0.0.1:8056/app/public/rest-api-email/v0/MailAccounts/cn_localmailaccount-1/MailFolders?_dc=1685500140332 net::ERR_FAILED

6.I put frontend and backend into same folder as a single PHP website(host http://127.0.0.1:8066),
and copy `desktop/resources/conjoon.conf.json` to `desktop/resources/conjoon.prod.conf.json`
and visit the website, and still can not save, and there is only error left:
GET http://127.0.0.1:8066/app/public/rest-api-email/v0/MailAccounts/cn_localmailaccount-1/MailFolders?_dc=1685501122622 404 (Not Found)

seems I did not set the url correctly...could you help ?

**Reply by ThorstenSuckow** [reply]
It looks like you have to configure your ISS to rewrite the URLs - take a look at how its done with `nginx`:

https://www.conjoon.org/docs/api/backends/@conjoon/lumen-app-email/#nginx-configuration

Since `lumen-app-email` is a **laravel** application, I have searched google and found this post: https://laravel-tricks.com/tricks/rewrite-urls-in-laravel-on-windows-with-iis

What you basically need to do is to rewrite each incoming url to the `index.php` script, which serves as the front controller for **lumen-app-email**.

Let me know if that helped!

**Reply by SupernovaCooper** [reply]
Hi,
Finally, it works~!
Thank you again.

At beginning, I thought only frontend is enought for me, so I just put frontend into IIS...
After asked you, I reallized that I need backend also, but I thought frontend and backend need to be two host(two websites? what should I call it ?)...So I got some problems...
And now I realized, actually it's should be in one single website.
Just as you mentioned:

doc_root/
   conjoon(frontend)/ <- conjoon goes here
   htdocs(backend)/ <- lumen-app-email goes here

And then edit the config file :`desktop/resources/conjoon.conf.json`
`extjs-app-localmailaccount` enable set to true,
`extjs-app-imapuser` enable set to false,
`extjs-app-webmail` `rest-api-email` set to "http://127.0.0.1:8066/htdocs/app/public/rest-api-email/v0/"

And the `.env` file for the backend, set `APP_URL=http://127.0.0.1:8066/htdocs/app/public`

And then everything should works, but in my case, I use phpEnv to run the server, so I followed this "https://www.conjoon.org/docs/api/backends/@conjoon/lumen-app-email/#nginx-configuration" add rewrite rules:
location /htdocs {
    try_files $uri $uri/ /htdocs/app/public/index.php?$query_string;
}
Now everything works, I can add my email accounts and checking my new emails.

Really appricate for everything...Thank you.

But I still have some problem:
1.When I click save button to save new account, seems nothing happened, no tips no dialogs, but after I refresh the browser then I can receive emails, I am not sure if I still did something wrong...
2.There is still error from the console: not fond `conjoon.prod.conf.json`.
3.how to set username and password for the system ? because now if I visit http://xxxx.com:xxx/conjoon all my accounts and emails shows up without any authentication...Because I want to checking my emails from other place, and I really don't want to set all the different client(android, iOS, PC and other port, because I have too much email account)...

Thank you.

**Reply by ThorstenSuckow** [reply]
Happy to hear that it worked out for you. 

So what you did was to configure conjoon with an auth_adapter that allows for multiple mail-accounts, but the mail-accounts are stored locally on your computer.

If you want to use conjoon on different computers without locally stored email-account-configuration, you can configure it for authenticating directly against an IMAP-account - for this, you'd have to revert a few steps:

## 1. disable `extjs-app-localmailaccount`, and enable `extjs-app-imapuser`:

```
 "extjs-app-localmailaccount": {
     "enabled": false,
...
 "extjs-app-imapuser": {
     "enabled": true,
...
```

## 2. Configure the endpoint for extjs-app-imapuser
[Configure `extjs-app-imapuser`](https://www.conjoon.org/docs/api/packages/@conjoon/extjs-app-imapuser#what-goes-into-an-extjs-app-imapuser-configuration) with the url where lumen-app-email-services are found: In this case, the path to `rest-imapuser` is required

```
"service": {
    "rest-imapuser" :  {
      "base" : ""http://127.0.0.1:8066/htdocs/app/public/rest-imapuser/v0/"
    }
  }
```

## 3. Configure AUTH_PROVIDER of lumen-app-email

set the `AUTH_PROVIDER` in the `lumen-app-email`-env configuration to `single-imap-user`:

```
APP_URL=https://your.iss-ip-address-or-domain-name.com/lumen-app-email-dir/app/public
APP_EMAIL_PATH=rest-api-email
AUTH_PROVIDER=single-imap-user
APP_ENV=production
APP_DEBUG=true
```

Refresh your browser and conjoon should show a login-mask - you'd then sign in with the login information as provided by your IMAP-email-account. There is a list of pre-configured email-services coming with lumen-app-email, if you're using anything else you'd have to configure the list of imap-servers you'd like to authenticate against:

See https://www.conjoon.org/docs/api/backends/@conjoon/lumen-app-email#imap-server-configurations

**Reply by SupernovaCooper** [reply]
Hi,
Thanks for reply.

I tried with your steps, but seems I can not login...
I tried to login with my email account and password, when I click login got error:
POST http://127.0.0.1:8066/conjoon/desktop/resources/extjs-lib-comp/html/blank.html 405 (Not Allowed)
POST http://127.0.0.1:8066/htdocs/app/public/rest-imapuser/v0/auth 401 (Unauthorized)


And seems `extjs-app-imapuser` need me to login with my email account and email password, right?
If right, maybe there are problems: 
1.All my email accounts enabled apps password(Two-factor authentication enabled, then seems can not login by main password anymore), but I can not remember the apps password at all (because the apps password is random and meaningless).
2.And another problem is, I don't want provide webmail service to anyone, I mean I don't want anyone login into my webmail system expect myself.

So, what I need is just a login page, login with the username and password that I have set, after login then I can check all my email accounts that I had set... seems like I need `extjs-app-localmailaccount` and a new login page and settings.
Is it possible ?

And ` but the mail-accounts are stored locally on your computer` means all my email accounts saved on the server locally or browser locally, if on the server side, then that is exactlly what I want...

Thank you.

**Reply by ThorstenSuckow** [reply]
> And seems `extjs-app-imapuser` need me to login with my email account and email password, right? If right, maybe there are problems: 1.All my email accounts enabled apps password(Two-factor authentication enabled, then seems can not login by main password anymore), but I can not remember the apps password at all (because the apps password is random and meaningless).

Yes, conjoon provides authentication means that require a one-time password for some accounts - it is further [described here](https://www.conjoon.org/docs/troubleshooting#q-when-do-i-have-to-use-a-one-time-password-with-my-free-email-account).

> 2.And another problem is, I don't want provide webmail service to anyone, I mean I don't want anyone login into my webmail system expect myself.

The default regexes consider a domain and then grant access to each address used with this domain. You can adjust the regex in the `imapserver.php` for each account so that it just works with the email addresses matching the regular expression. See [this part of the documentation](https://www.conjoon.org/docs/api/backends/@conjoon/lumen-app-email#example-for-match) for more information.

> So, what I need is just a login page, login with the username and password that I have set, after login then I can check all my email accounts that I had set... seems like I need `extjs-app-localmailaccount` and a new login page and settings. Is it possible ?
> 
> And ` but the mail-accounts are stored locally on your computer` means all my email accounts saved on the server locally or browser locally, if on the server side, then that is exactlly what I want...

There is no authentication handler that allows for persisting email accounts server side yet, since this would mean additional coupling with infrastructure which development did not focus on for V1.*. So the only options are right now

1. Signing in using the IMAP account information 
2. Storing the mail accounts locally on the computer

It should be doable providing a quick hack that serves `extjs-app-localmailaccount`-functionality with an existing backend (instead of HTML5 Local Storage), I just don't know how quick I'll be able to whip this up...

**Reply by SupernovaCooper** [reply]
Thank you so much.
